lib/src/bitarray.dart

```diff
@@ -0,0 +1,317 @@
+    RangeError.checkValidIndex(index, this);
+    return (_data[index >> 5] & (1 << (index & 31))) != 0;
+  }
+
+  @override
+  int get length => _length;
+
+  @override
+  void operator []=(int index, bool val) {
+    RangeError.checkValidIndex(index, this);
+    if (val) {
+      _data[index >> 5] |= 1 << (index & 31);
+    } else {
+      _data[index >> 5] &= ~(1 << (index & 31));
+    }
+  }
+
+  /// Counts number of bits in unsigned [val].
+  int _countBits(int val) {
+    var bits = 0;
+    while (val != 0) {
+      bits += _numOfBitsLookup[val & 0xFF];
+      val >>= 8;
+    }
+    return bits;
+  }
+
+  /// Returns number [true] values in [this].
+  int get cardinality {
+    var card = 0;
+    for (var word in _data) {
+      card += _countBits(word);
+    }
+    return card;
+  }
+
+  /// Returns logical NOT of [this].
+  BitArray operator ~() => BitArray.notGate(this);
+
+  /// Returns result or logical AND between [this] and [ba].
+  BitArray operator &(BitArray ba) => BitArray.andGate(this, ba);
+
+  /// Returns result or logical OR between [this] and [ba].
+  BitArray operator |(BitArray ba) => BitArray.orGate(this, ba);
+
+  /// Returns result or logical XOR between [this] and [ba].
+  BitArray operator ^(BitArray ba) => BitArray.xorGate(this, ba);
+
+  /// Checks lengths equal for given [BitArray]s.
+  ///
+  /// Throws [ArgumentExceprion] when lengts is differ.
+  static void _checkLengthsEqual(BitArray ba1, BitArray ba2) {
+    if (ba1._length != ba2._length) {
+      throw ArgumentError.value(
+          'BitArrays should have equal lengths: ${ba1._length} != ${ba2._length}');
+    }
+  }
+
+  /// Sets the bit at the [index] to the complement of its current value.
+  void flip(int index) {
+    RangeError.checkValidIndex(index, this);
+    _data[index >> 5] ^= 1 << (index & 31);
+  }
+
+  /// Sets each bit at the range from [start] inclusive to [end] exclusive to the complement of its current value.
+  void flipRange(int start, [int end]) {
+    end ??= _length;
+    RangeError.checkValidRange(start, end, _length);
+
+    var startWordIndex = start >> 5;
+    var endWordIndex = end >> 5;
+
+    if (startWordIndex == endWordIndex) {
+      for (var pos = start; pos < end; ++pos) {
+        flip(pos);
+      }
+    } else {
+      // First word
+      _data[startWordIndex] ^= -1 << (start & 31);
+
+      // Intermediate words
+      for (var wordIndex = startWordIndex + 1;
+          wordIndex < endWordIndex;
+          ++wordIndex) {
+        _data[wordIndex] = ~_data[wordIndex];
+      }
+
+      // Last word
+      _data[endWordIndex] ^= 0xFFFFFFFF >> (32 - (end & 31));
+    }
+  }
+
+  @override
+  Iterator<bool> get iterator => BitArrayIterator(this);
+}
+
+/// Iterator for [BitArray].
+class BitArrayIterator implements Iterator<bool> {
+  bool _current;
+  int _pos = 0;
+  final BitArray _bitArray;
+
+  BitArrayIterator(this._bitArray);
+
+  @override
+  bool get current => _current;
+
+  @override
+  bool moveNext() {
```

Comment:
If you want a custom iterator anyway, consider preparing for Null Safety by specifying that the iterator returns `false` before calling `moveNext` or after `moveNext` has returned false, instead of `null`.
(This should be specified on the `iterator` getter of `BitArray` if this class is private).

Then initialize `_current` to `false` and set it to `false` instead of `null` below.

Reply:
Is using Null Safety OK for custom iterators? Returning non-`null` values before and after iteration violates requirements of `current` property of `Iterator<E>`:
> Returns null if the iterator has not yet been moved to the first element, or if the iterator has been moved past the last element of the Iterable.